### DISCUSSION TITLE
Is it possible to create custom widgets like highcharts for 2D/3D data visualization?

### DISCUSSION THREAD

**Discussion by cokkiy** [slint-ui/slint#3350] [Ideas]
Is it possible to create custom widgets like highcharts for 2D/3D data visualization? I tried to find a widget for data visualization but had no results and no document mentions how to create customer widgets, Can the developer create a customer widget anyway? Or does any roadmap allow the dev to extend the slint?

**Comment by ogoffart** [comment]
Something like this ?  https://slint.dev/demos/plotter/

To create such a widget you would typically do something like this:

```
export global MyWidgetLogic {
   pure callback render(width: length, height: length, some_value: string, some_other_value: int) -> image;
}

export component MyWidget {
   in property <string> some_value;
   private property <int> some_other_value;
   Image {
      source: MyWidgetLogic.render(root.width, root.height, some_value, some_other_value);
   }
   TouchArea {
       pointer-event => { /* .. do something, maybr change some_other_value .. */ }
   }
}
```

I hope this helps.

If render can be slow, you can, in your logic, cache the value. And if it is not in cache, return a placeholder image. And depend on another property such that you can render in a thread and set that property when it is ready.
